### ISSUE TITLE
[Metadata Immutability]

### ISSUE THREAD

**Issue by sandeshkr419** [opensearch-project/OpenSearch#8647]
**Is your feature request related to a problem? Please describe.**
[Metadata](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/metadata/Metadata.java) object as part of [ClusterState](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/ClusterState.java) is supposed to be immutable by design. However, certain objects within Metadata do not reflect this design.

The following arrays should be replaced with lists and should implement `Collections.unmodifiableList()` or any other immutable implementations. Assigning them as arrays leave them open for mutability.
https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/metadata/Metadata.java#L268 
```
 String[] allIndices,
 String[] visibleIndices,
 String[] allOpenIndices,
 String[] visibleOpenIndices,
 String[] allClosedIndices,
 String[] visibleClosedIndices,
```

Also, the usage of these arrays is prominently in [IndexNameExpressionResolver](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/metadata/IndexNameExpressionResolver.java#L1175) where before these arrays are utilized, they are converted to lists before they are consumed.
It is thus unreasonable to keep these objects as arrays as opposed to lists.

**Describe the solution you'd like**
Store the above objects as lists, and implement them as `Collections.unmodifiableList()`.

**Describe alternatives you've considered**
Briefly discussed this change [here](https://github.com/opensearch-project/OpenSearch/pull/7853/files/e347287ab20a3f19e306d666137774046c35b348#r1230411484), but this change was out of scope of https://github.com/opensearch-project/OpenSearch/pull/7853

**Additional context**
this change will require changes in how the metadata or its diff is parsed as XContent as well.

**Comment by andrross** [comment]
What do you think about [this comment](https://github.com/opensearch-project/OpenSearch/blob/e3542011cd9584c354405ba6cbf9a31ced7bc5ce/server/src/main/java/org/opensearch/cluster/metadata/Metadata.java#L1559-L1562):

```
// TODO: I think we can remove these arrays. it isn't worth the effort, for operations on all indices.
// When doing an operation across all indices, most of the time is spent on actually going to all shards and
// do the required operations, the bottleneck isn't resolving expressions into concrete indices.
```

While I completely agree with you that immutable lists are better than mutable arrays, my first thought when digging into this code was to remove the arrays all together, then I found the above comment. If I understand correctly, then the existing getters would change to compute the results on demand. For example, `getConcreteVisibleIndices()` would change to:

```java
/**
 * Returns all the concrete indices that are not hidden.
 */
public List<String> getConcreteVisibleIndices() {
    return indices.entrySet().stream()
        .filter(e -> IndexMetadata.INDEX_HIDDEN_SETTING.get(e.getValue().getSettings()) == false)
        .map(Map.Entry::getKey)
        .collect(Collectors.toUnmodifiableList());
}
```

(This also changes the return type from String[] to List<String>, which isn't strictly necessary since mutability is no longer a concern given that a new array would be computed each invocation. However, if this is always being used as a list then it should probably just return a list.)

Pre-computing the filtered results strike me as an unnecessary optimization and I would suggest doing the simpler thing unless we know for sure there are performance benefits to pre-computing.

**Comment by sandeshkr419** [comment]
@andrross Thanks for taking a look. I had visited the [comment](https://github.com/opensearch-project/OpenSearch/blob/e3542011cd9584c354405ba6cbf9a31ced7bc5ce/server/src/main/java/org/opensearch/cluster/metadata/Metadata.java#L1559-L1562) you mentioned in my previous PR https://github.com/opensearch-project/OpenSearch/pull/7853 which touched the same code base.

I don't support the idea of computing different indices lookups (arrays in concern) on demand is that these different different arrays are used for expression resolving which is invoked in each almost all search and index data modification related calls. Calculating these on demand will surely increase the latency of all search and indexing related operations, which we should avoid. 
Computing these arrays with each cluster state update does not even takes time of the order of miliseconds. Plus trading off some memory/time in ClusterState update activity seems a better choice than to put it on indexing/search tasks where latency is required at minimal.

Also, I would like to point out that except the `getConcreteAllIndices()`, all the other arrays (5 of them) are only invoked at this [resolveEmptyOrTrivialWildcard()](https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/cluster/metadata/IndexNameExpressionResolver.java#L1175) in the entire code base, where they are used as lists directly.

**Comment by andrross** [comment]
Thanks @sandeshkr419! It makes total sense if these data structures are used in hot path indexing and search code then precomputing can be a good tradeoff. However, these TODO comments should be resolved or removed to avoid future confusion. The TODO comment referenced in #7853 actually does make a lot of sense (if I understand it correctly) as it is suggests to move the pre-computed structures into IndexNameExpressionResolver. That would likely make the code and intent more clear as the structures would more clearly be in hot path code (index name resolution). I haven't looked into what it would take to do that refactoring. Either way, I suggest addressing or removing the TODOs wherever possible.

**Comment by sandeshkr419** [comment]
@andrross Yes, we should remove these comments altogether. We did not removed it previously just to be sure that we have not misunderstood the correct context of these comments. Plus, there is also discussion which seems to be initiated in future by @shwetathareja on remote cluster state as in https://github.com/opensearch-project/OpenSearch/issues/7002#issuecomment-1573277658

For this issue, I would like to propose a minor enhancement only in terms of Metadata immutability.

**Comment by sandeshkr419** [comment]
@andrross Seems reasonable to remove the comment since we have valid reasons (potential performance regressions) of not implementing this.

@shwetathareja What are your thoughts on this (removing the TODOs - since we have reason of not implementing them).

**Comment by shwetathareja** [comment]
Like the comment said, these arrays are used during index resolution. We can benchmark to see how many cpu cycles are spent to generate these arrays during ingestion and search traffic. Personally I think even if there is small performance gain, we should keep them. @sandeshkr419 @andrross

**Comment by andrross** [comment]
@sandeshkr419 @shwetathareja Just to summarize our discussion, I think we're in agreement we should keep the pre-computation. So that means we should remove the [TODO comment](https://github.com/opensearch-project/OpenSearch/blob/e3542011cd9584c354405ba6cbf9a31ced7bc5ce/server/src/main/java/org/opensearch/cluster/metadata/Metadata.java#L1560-L1562). I'm also on board with changing the arrays to immutable List instances. And finally, I think we should probably add additional versions of the getters (we need to keep the existing signatures for compatibility) that return List (or Collection) types because a quick glance around the code base shows that many times these arrays are just immediately converted to lists.

I'm also going to tag this with the "good first issue" as this looks like it should be a fairly straightforward refactoring. Let me know if you think this is wrong or have anything else to add. Thanks!

**Comment by akolarkunnu** [comment]
@sandeshkr419 @andrross 
Can you please assign this bug to me, if no one is working on this?

**Comment by akolarkunnu** [comment]
@sandeshkr419 @andrross
Why do we need to keep two versions of all get APIs(Eg: getConcreteVisibleIndices()) , one returns array of Strings and new one returns List ? We can change all call places within the OpenSearch code base along with this change. Are you expecting that these APIs might have used from out of OpenSerach code base?

If yes, shall we deprecate old API(which returns String []) along with this change and introduce new API. 
A note, the new API name will be different from old one, since we can't have overridden APIs which have difference only in return types.

**Comment by andrross** [comment]
> Are you expecting that these APIs might have used from out of OpenSerach code base?

@akolarkunnu Yes. This class is exposed to all OpenSearch plugins.
